File: exam-system-java/server/src/main/java/com/yyjy/exam/question/repository/CategoriesRepository.java
package com.yyjy.exam.question.repository;

import com.yyjy.exam.entity.question.dto.QuestionsCategoriesTree;
import com.yyjy.exam.entity.question.entity.QuestionsCategories;
import com.yyjy.exam.entity.question.entity.QuestionsCategoriesTable;
import com.yyjy.exam.entity.question.entity.QuestionsTable;
import org.babyfish.jimmer.spring.repository.JRepository;
import org.babyfish.jimmer.sql.fetcher.Fetcher;

import java.util.List;

public interface CategoriesRepository extends JRepository<QuestionsCategories, Long> {
	
	List<QuestionsCategories> findByParentId(long parentId);
	
	boolean existsByParentId(long parentId);
	
	boolean existsByParentIdAndName(Long parentId, String name);
	
	default List<QuestionsCategoriesTree> findTree() {
		var t = QuestionsCategoriesTable.$;
		List<QuestionsCategoriesTree> items = sql().createQuery(t)
				                                      .where(t.parentId().eq(0L))
				                                      .orderBy(t.sort())
				                                      .select(t.fetch(QuestionsCategoriesTree.class))
				                                      .execute();
		calculateCount(items);
		return items;
	}
	
	default int calculateCount(List<QuestionsCategoriesTree> categories) {
		int total = 0;
		for (var node : categories) {
			List<QuestionsCategoriesTree> children = node.getChildren();
			if (children == null || children.isEmpty()) {
				total += node.getQuestionCount();
			} else {
				int childrenSum = calculateCount(children);
				node.setQuestionCount(childrenSum);
				total += childrenSum;
			}
		}
		return total;
	}
	
	default boolean hasQuestions(Long categoryId) {
		QuestionsTable qt = QuestionsTable.$;
		return !sql().createQuery(qt)
				        .where(qt.categoryId().eq(categoryId))
				        .select(qt)
				        .limit(1)
				        .execute()
				        .isEmpty();
	}
	
	default List<QuestionsCategories> findAllWithFetcher(Fetcher<QuestionsCategories> fetcher) {
		return sql().getEntities().findAll(fetcher);
	}
}

File: exam-system-server/server/src/main/kotlin/com/yyjy/repository/CategoriesRepository.kt
package com.yyjy.repository

import com.yyjy.models.entity.Categories
import com.yyjy.models.entity.dto.CategoriesTree
import com.yyjy.models.entity.parentId
import com.yyjy.models.entity.sort
import org.babyfish.jimmer.spring.repository.KRepository
import org.babyfish.jimmer.sql.kt.ast.expression.eq

/**
 * @author Nyxcirea
 * date 2026/4/1
 * description: TODO
 */
interface CategoriesRepository : KRepository<Categories, Long> {
    fun findTree(): List<CategoriesTree> {
        val items = sql.createQuery(Categories::class) {
            where(table.parentId eq 0)
            orderBy(table.sort)
            select(table.fetch(CategoriesTree::class))
        }.execute()
        calculateCount(items)

        return items
    }

    fun calculateCount(categories: List<CategoriesTree>): Int {
        var total = 0
        for (node in categories) {
            if (node.children!!.isEmpty()) {
                // 如果是叶子节点，保持它原有的 count (或者根据业务逻辑处理)
                total += node.questionCount
            } else {
                // 如果有子节点，递归计算子节点的总和
                val childrenSum = node.children?.let { calculateCount(it) }
                if (childrenSum != null) {
                    node.questionCount = childrenSum
                    // 将汇总值赋给父节点
                    total += childrenSum
                }
            }
        }
        return total
    }

    fun existsByParentIdAndName(parentId: Long, name: String): Boolean
    fun existsByParentId(parentId: Long): Boolean
}
